### ISSUE TITLE
Невозможно описать шаблонный делегат, если типы его параметров/результата требуют ограничений шаблона

### ISSUE THREAD

**Issue by SunSerega** [pascalabcnet/pascalabcnet#2722]
```
type
  TLimited<T> = class
    where T: record;
  end;
  
  //Ошибка: Невозможно инстанцировать, так как тип T не является размерным
  TMyFunc<T> = function(a: TLimited<T>): T?;
  
begin end.
```

Я вижу 2 основных варианта решения:
1. Разрешить `where` после описания типа-делегата.
Более красивое решение, но требует изменения синтаксиса.

2. Проверять ограничения шаблонов там, где используется `TMyFunc<T>`.
Ошибку может быть сложнее понять ("Почему в `TMyFunc<T>` нельзя подставить такой то конкретный `<T>`?"). Но с синонимами типов та же ситуация.
И из того что я по быстрому попытался набросать - типы параметров и результата инстанцирует ещё когда превращает `TMyFunc` из синтаксического узла в семантический. Поэтому если не проверять сразу - компилятор крашится при попытке получить рефлекцию типа `T?`, не указав явно что `T`-запись. (потому что `Nullable` это тип из .dll)

В принципе можно ещё выставлять ограничения на шаблон типа-делегата автоматически. Но это и с остальным языком не сочитается, и потребует довольно много нового кода в компиляторе.

**Comment by ibond84** [comment]
Разрешить where после описания типа-делегата.

**Comment by ibond84** [comment]
В C# можно накладывать ограничения на генерик-параметры делегатов.

**Comment by ibond84** [comment]
>Отложить проверку ограничений шаблонов. Проверять её там, где используется TMyFunc<T>.

Вы путаете шаблоны с генериками..

**Comment by SunSerega** [comment]
> Вы путаете шаблоны с генериками..

Нет, я прекрасно понимаю разницу. Но я вам указал на то, что вы уже сами их спутали:
```
type
  Generic<T1,T2> = class
    where T1: record;
    where T2: class;
  end;
  
  Synonym<T> = Generic<Exception, byte>;
  
begin
  var x: Synonym<word>;
end.
```
В этом коде ошибка показывает на `word` и говорит:

> Невозможно инстанцировать, так как тип System.Exception не является размерным

При этом если убрать объявление `x` - ошибку вообще не показывает, хотя проблема в объявлении `Synonym`.
Ну и это простой пример, у меня в `OpenCLABC` в паре мест шаблонные параметры аналога `Synonym` передаются в аналог `Generic`, и только определённые комбинации являются допустимимы. Но я никак не могу указать это в объявлении `Synonym`. Поэтому получаются кривые ошибки.

---

Ну, разрешить писать `where` таки было бы лучшим решением. Но может тогда сразу так же требовать `where` после синонимов?
